Cython/Utility/ObjectHandling.c

```diff
@@ -1872,6 +1914,37 @@ static PyObject* __Pyx__CallUnboundCMethod0(__Pyx_CachedCFunction* cfunc, PyObje
 //@requires: UnpackUnboundCMethod
 //@requires: PyObjectCallOneArg
 
+#if CYTHON_COMPILING_IN_CPYTHON
+// FASTCALL methods receive "&empty_tuple" as simple "PyObject[0]*"
+static CYTHON_INLINE PyObject* __Pyx_CallUnboundCMethod0(__Pyx_CachedCFunction* cfunc, PyObject* self) {
+    int was_initialized = __Pyx_CachedCFunction_GetAndSetInitializing(cfunc);
+
+    if (likely(was_initialized==2 && cfunc->func)) {
+        return (likely(cfunc->flag == METH_NOARGS) ?  (*(cfunc->func))(self, NULL) :
+            (likely(cfunc->flag == METH_FASTCALL) ?
+                (*(__Pyx_PyCFunctionFast)(void*)(PyCFunction)cfunc->func)(self, &EMPTY(tuple), 0) :
+            ((cfunc)->flag == (METH_FASTCALL | METH_KEYWORDS) ?
+                (*(__Pyx_PyCFunctionFastWithKeywords)(void*)(PyCFunction)cfunc->func)(self, &EMPTY(tuple), 0, NULL) :
+                (likely(cfunc->flag == (METH_VARARGS | METH_KEYWORDS)) ?  ((*(PyCFunctionWithKeywords)(void*)(PyCFunction)cfunc->func)(self, EMPTY(tuple), NULL)) :
+                (cfunc->flag == METH_VARARGS ?  (*(cfunc->func))(self, EMPTY(tuple)) :
+                __Pyx__CallUnboundCMethod0(cfunc, self))))));
+    }
```

Comment:
This style is high limit for a C macro, but an inline function should really use separate if-statements (or even `switch(cfunc->flag)`) to make it clearer what the conditions are and what happens here.

Reply:
I think this was a conscious effort not to change it too much. But I've unpicked it now.